File: src/ypl/merge.py
"""The three-way merge behind a reconcile.

Pure: lists of video ids in, lists of video ids out. Nothing here reads a file
or touches YouTube, because this is the part that has to be reasoned about
exactly and a test that has to build a playlist file to state a case is a test
nobody writes.

Three inputs, and the third is the point. Local `[A, B, C]` against remote
`[A, C]` has two possible histories — B was deleted on the phone, or B was
added here and never pushed — and they are the same two lists with opposite
correct actions. The base, what YouTube held at the last reconcile, is what
tells them apart: anything in the base and missing from a side was deleted by
that side, and anything on a side and missing from the base was added by it.

Membership merges per video. Order does not: it is settled once for the whole
playlist, because per-item order merging turns a moved track into an argument
with no answer, and a playlist whose order is half one machine's and half
another's is worse than either.
"""

from dataclasses import dataclass
from dataclasses import field

LOCAL = 'local'
REMOTE = 'remote'

# One occurrence of one video: the id, and which copy of it this is. A playlist
# may legitimately hold the same mix twice — YouTube allows it and ypl adds
# with `duplicates=True` — so a bare set of ids would read the second copy as
# noise and quietly drop it. Keying by occurrence turns counting into set
# arithmetic and keeps every copy accounted for.
Key = tuple[str, int]


@dataclass
class Merge:
    """What the reconcile decided, and what is left to push.

    The pending lists are reporting rather than a queue. What has to go up is
    always re-derivable by comparing the local file against the base, which is
    why the queue is state rather than data and why nothing here is written
    down as an instruction to be replayed later.
    """

    order: list[str] = field(default_factory=list)
    pulled_in: list[str] = field(default_factory=list)
    pulled_out: list[str] = field(default_factory=list)
    pending_add: list[str] = field(default_factory=list)
    pending_remove: list[str] = field(default_factory=list)
    order_source: str = LOCAL

    @property
    def changed_here(self) -> bool:
        return bool(self.pulled_in or self.pulled_out)

    @property
    def to_push(self) -> bool:
        return bool(self.pending_add or self.pending_remove)


def keyed(video_ids: list[str]) -> list[Key]:
    """Number each video by which copy of itself it is, in order."""
    seen: dict[str, int] = {}
    keys = []
    for video_id in video_ids:
        keys.append((video_id, seen.get(video_id, 0)))
        seen[video_id] = seen.get(video_id, 0) + 1
    return keys


def reordered(base: list[Key], other: list[Key]) -> bool:
    """Whether `other` moved anything relative to the base.

    Only the videos both hold can answer this. Comparing the two lists whole
    would call every addition and every removal a reorder, which would hand
    remote the ordering of the entire playlist every time anyone added a track
    on their phone.
    """
    common = set(base) & set(other)
    return [key for key in base if key in common] != [key for key in other if key in common]


def weave(primary: list[Key], secondary: list[Key], wanted: set[Key]) -> list[Key]:
    """Order `wanted` by `primary`, slotting in what only `secondary` knows.

    An item the winning order has never seen goes where the other order put it
    — directly after whichever item precedes it there — rather than at the end.
    Appending would be simpler and would move a track added at the front of the
    playlist on a phone to the back of it here, which reads as ypl having
    reordered something nobody touched.

    The cursor only ever moves forward. Without that, two orders that disagree
    about the items they share drag it backwards: local `[c, b, a]` against
    remote `[a, b, c, g]` would anchor g to c, the last thing remote saw before
    it, and land it second in a playlist where remote meant it last. Held
    forward, an anchor that has stopped meaning anything degrades to the end of
    the list, which is where the side that reordered had it.
    """
    result = [key for key in primary if key in wanted]
    cursor = -1
    for key in secondary:
        if key not in wanted:
            continue
        if key in result:
            cursor = max(cursor, result.index(key))
            continue
        cursor += 1
        result.insert(cursor, key)
    return result


def merge(base: list[str], remote: list[str], local: list[str]) -> Merge:
    """Reconcile one playlist's three states.

    An empty base means this playlist has never been reconciled, and the result
    is the union of both sides: with nothing recorded, no absence can be read
    as a deletion, and guessing wrong would delete a playlist's worth of videos
    on the strength of never having looked.
    """
    base_keys, remote_keys, local_keys = keyed(base), keyed(remote), keyed(local)
    base_set, remote_set, local_set = set(base_keys), set(remote_keys), set(local_keys)

    removed_remotely = base_set - remote_set
    removed_locally = base_set - local_set
    wanted = (remote_set | local_set) - removed_remotely - removed_locally

    # Remote wins, so the only question is whether remote reordered at all. A
    # local reorder survives only while remote left its own order alone.
    order_source = REMOTE if reordered(base_keys, remote_keys) else LOCAL
    primary, secondary = (remote_keys, local_keys) if order_source == REMOTE else (local_keys, remote_keys)

    # A video both sides added independently is in neither report: it needs
    # nothing done to it, and pushing it again would make it a duplicate.
    arrived = wanted - base_set - local_set
    departed = removed_remotely & local_set
    to_add = wanted - base_set - remote_set
    to_remove = removed_locally & remote_set

    return Merge(
        order=[video_id for video_id, _ in weave(primary, secondary, wanted)],
        pulled_in=[video_id for video_id, _ in (key for key in remote_keys if key in arrived)],
        pulled_out=[video_id for video_id, _ in (key for key in base_keys if key in departed)],
        pending_add=[video_id for video_id, _ in (key for key in local_keys if key in to_add)],
        pending_remove=[video_id for video_id, _ in (key for key in base_keys if key in to_remove)],
        order_source=order_source,
    )

File: api/merge/merge.go
// Package merge decides what syncing one playlist does: how YouTube's read of it
// merges into the server's order, and which writes make YouTube hold the
// result. It reads no store and makes no request.
//
// A merge has three inputs. The base is what YouTube held after the server last
// read the playlist, with each write the server made since, the read is what
// YouTube holds now, and the entries are the server's order. Every one of them
// is keyed by YouTube's playlistItem id, which a copy of a video keeps wherever
// it moves. The base is what gives the other two meaning:
//
//   - an item of the base the read lacks was removed on YouTube
//   - an item of the base no entry holds was removed here
//   - an item of the read the base lacks was added on YouTube
//   - an entry holding no item was added here
//
// Order is settled once for the whole playlist, because a playlist ordered half
// by one side and half by the other is worse than either order.
package merge

import (
	"fmt"
	"slices"
)

// Item is one item of a playlist on YouTube: its playlistItem id and the video
// in it.
type Item struct {
	ID      string
	VideoID string
}

// BaseItem is one item of a playlist's base, and whether its position is where
// a write the server made put it, which no read has shown. YouTube holds such an
// item where the write landed, and an edit on YouTube while the write was sent
// shifts where that is.
type BaseItem struct {
	Item
	Placed bool
}

// Entry is one entry of the server's order of a playlist: its id, the video in
// it, and the id of the item holding the video on YouTube, which is empty for an
// entry added here and not yet pushed. An entry whose ID is 0 is new.
type Entry struct {
	ID      int64
	VideoID string
	ItemID  string
}

// Side names whose order a merge kept.
type Side string

const (
	// Server is the server's order.
	Server Side = "server"
	// YouTube is YouTube's order.
	YouTube Side = "youtube"
)

// Sort is how YouTube orders a playlist.
type Sort string

const (
	// Manual is a playlist YouTube keeps in the order writes put it in.
	Manual Sort = "manual"
	// Automatic is a playlist YouTube orders itself, which refuses a write
	// naming a position.
	Automatic Sort = "automatic"
)

// Playlist is what a merge of one playlist reads.
type Playlist struct {
	// Base is what YouTube held after the server last read the playlist, with
	// each write the server made since.
	Base []BaseItem
	// Read is what YouTube holds now. Every item of Base the read lacks has to
	// have been confirmed gone before the merge, since a read that spans pages
	// can miss an item that is there.
	Read []Item
	// Entries is the server's order.
	Entries []Entry
	// Unanswered is the write sent against Base whose answer was never
	// recorded, so YouTube may hold it too, and nil when there is none.
	Unanswered *Write
	// Sort is how YouTube orders the playlist.
	Sort Sort
}

// Result is what a merge decided.
type Result struct {
	// Entries is the server's order after the merge.
	Entries []Entry
	// Order is the side whose order Entries keeps.
	Order Side
	// Added is how many items YouTube added that the merge gave an entry, and
	// Removed how many entries the merge dropped because YouTube removed their
	// item.
	Added, Removed int
	// Changed is whether Entries differs from the entries merged.
	Changed bool
}

// Merge merges the read of a playlist into its entries against its base.
//
// YouTube's order wins when YouTube orders the playlist itself, and when it
// moved an item it placed itself among those the base and the read share. The
// server's order wins otherwise.
//
// An insert whose answer was lost may have made an item. The first item of the
// read holding its video that the base lacks and no entry holds is taken for
// it: the entry the insert was sent for holds it, or, when an edit here removed
// that entry, the item is removed here. So an item YouTube added with the same
// video before the read is taken for the insert, which is one edit a merge
// loses. A move on YouTube of an item the server placed, before a read shows
// the placement, is the other.
//
// An entry holding an item neither the base nor the read has becomes an entry
// added here again, so its video is pushed rather than lost.
func Merge(p Playlist) (Result, error) {
	if p.Sort != Manual && p.Sort != Automatic {
		return Result{}, fmt.Errorf("merge a playlist sorted %q, which is not a sort a merge knows", p.Sort)
	}
	merging := slices.Clone(p.Entries)
	inBase, inRead := map[string]bool{}, itemSet(p.Read)
	for _, item := range p.Base {
		inBase[item.ID] = true
	}
	madeHere, err := landedInsert(p, merging)
	if err != nil {
		return Result{}, err
	}
	held := map[string]bool{}
	for i, entry := range merging {
		if entry.ItemID != "" && !inBase[entry.ItemID] && !inRead[entry.ItemID] {
			merging[i].ItemID = ""
		}
		if merging[i].ItemID != "" {
			held[merging[i].ItemID] = true
		}
	}

	// A key is an item's id, or a pending entry's place among the entries.
	type key struct {
		item    string
		pending int
	}
	entryKey := func(i int) key {
		if merging[i].ItemID == "" {
			return key{pending: i + 1}
		}
		return key{item: merging[i].ItemID}
	}
	wanted := map[key]bool{}
	var serverKeys, readKeys []key
	result := Result{Order: Server}
	for i, entry := range merging {
		k := entryKey(i)
		serverKeys = append(serverKeys, k)
		if entry.ItemID != "" && inBase[entry.ItemID] && !inRead[entry.ItemID] {
			result.Removed++
			continue
		}
		wanted[k] = true
	}
	videoOf := map[string]string{}
	for _, item := range p.Read {
		k := key{item: item.ID}
		readKeys = append(readKeys, k)
		videoOf[item.ID] = item.VideoID
		switch {
		case held[item.ID]:
			wanted[k] = true
		case inBase[item.ID], item.ID == madeHere:
		default:
			result.Added++
			wanted[k] = true
		}
	}

	primary, secondary := serverKeys, readKeys
	if p.Sort == Automatic || reordered(p) {
		result.Order = YouTube
		primary, secondary = readKeys, serverKeys
	}
	entryOf := map[string]Entry{}
	for _, entry := range merging {
		if entry.ItemID != "" {
			entryOf[entry.ItemID] = entry
		}
	}
	for _, k := range weave(primary, secondary, wanted) {
		switch {
		case k.pending > 0:
			result.Entries = append(result.Entries, merging[k.pending-1])
		case entryOf[k.item].ItemID != "":
			result.Entries = append(result.Entries, entryOf[k.item])
		default:
			result.Entries = append(result.Entries, Entry{VideoID: videoOf[k.item], ItemID: k.item})
		}
	}
	result.Changed = !slices.Equal(result.Entries, p.Entries)
	return result, nil
}

// landedInsert settles in entries the item an unanswered insert made, if the
// read shows one: the entry the insert was sent for takes it while that entry
// is still waiting for an item. It returns the item when no entry takes it,
// since an edit here removed the entry, and empty otherwise.
func landedInsert(p Playlist, entries []Entry) (string, error) {
	w := p.Unanswered
	if w == nil {
		return "", nil
	}
	switch w.Kind {
	case Insert, Append:
	case Delete, Move:
		return "", nil
	default:
		return "", fmt.Errorf("merge after an unanswered write of kind %q, which is not a kind a merge knows", w.Kind)
	}
	taken := map[string]bool{}
	for _, item := range p.Base {
		taken[item.ID] = true
	}
	for _, entry := range entries {
		if entry.ItemID != "" {
			taken[entry.ItemID] = true
		}
	}
	index := slices.IndexFunc(p.Read, func(item Item) bool { return item.VideoID == w.VideoID && !taken[item.ID] })
	if index < 0 {
		return "", nil
	}
	made := p.Read[index].ID
	for i, entry := range entries {
		if entry.ID == w.EntryID && entry.ItemID == "" {
			entries[i].ItemID = made
			return "", nil
		}
	}
	return made, nil
}

// reordered is whether YouTube moved an item it placed itself among those the
// base and the read share. Only those can answer it. Comparing whole lists would
// call every addition and removal a reorder. An item a write here placed sits
// where the write landed, which an edit on YouTube while it was sent shifts, and
// so does the item an unanswered move names.
func reordered(p Playlist) bool {
	unplaced := map[string]bool{}
	for _, item := range p.Base {
		unplaced[item.ID] = !item.Placed
	}
	if w := p.Unanswered; w != nil && w.Kind == Move {
		unplaced[w.ItemID] = false
	}
	inRead := itemSet(p.Read)
	var base, read []string
	for _, item := range p.Base {
		if unplaced[item.ID] && inRead[item.ID] {
			base = append(base, item.ID)
		}
	}
	for _, item := range p.Read {
		if unplaced[item.ID] {
			read = append(read, item.ID)
		}
	}
	return !slices.Equal(base, read)
}

// weave is the wanted keys of primary in its order, with each wanted key only
// secondary holds placed directly after the key that precedes it in secondary.
//
// The insertion point only moves forward. Where the two orders disagree about
// the keys they share, a key secondary holds after all of them lands at the end,
// where secondary had it, rather than beside an anchor primary placed earlier.
func weave[K comparable](primary, secondary []K, wanted map[K]bool) []K {
	result := slices.DeleteFunc(slices.Clone(primary), func(k K) bool { return !wanted[k] })
	cursor := -1
	for _, k := range secondary {
		if !wanted[k] {
			continue
		}
		if index := slices.Index(result, k); index >= 0 {
			cursor = max(cursor, index)
			continue
		}
		cursor++
		result = slices.Insert(result, cursor, k)
	}
	return result
}

func itemSet(items []Item) map[string]bool {
	set := make(map[string]bool, len(items))
	for _, item := range items {
		set[item.ID] = true
	}
	return set
}
